sirepo/pkcli/test_http.py

```diff
@@ -322,60 +288,230 @@ def _uri(self, uri):
     global cfg
     if cfg:
         return
     cfg = pkconfig.init(
         server_uri=('http://127.0.0.1:8000', str, 'where to send requests'),
+        run_min_secs=(90, int, 'minimum amount of time to let a simulation run'),
+        run_max_secs=(120, int, 'maximum amount of time to let a simulation run'),
     )
 
 
-async def _run(email, sim_type):
-    await _run_sequential_parallel(
-        await _Client(email=email, sim_type=sim_type).login(),
-    )
+async def _main():
+    async def _get_apps(clients):
+        a = []
+        for c in clients:
+            for t in (
+                    'elegant',
+                    'jspec',
+                    'srw',
+                    'synergia',
+                    'warppba',
+                    'warpvnd',
+            ):
+                a.append(await _App(t, c.copy()).setup_sim_data())
+        return a
+
+    async def _get_clients(*users):
+        l = []
+        for u in users:
+            l.append(_Client(u).login())
+        return await asyncio.gather(*l)
+
+    def _get_sims(apps):
+        s = []
+        for a in apps:
+            t = _CODES[a.sim_type]
+            e = t[random.randrange(len(t))]
+            random.shuffle(e.reports)
+            for r in e.reports:
+                s.append(_Sim(a, e.name, r).run())
+        return s
 
+    def _register_signal_handlers(main_task):
+        def _s(*args):
+            main_task.cancel()
+        signal.signal(signal.SIGTERM, _s)
+        signal.signal(signal.SIGINT, _s)
 
-async def _run_all():
-    l = []
-    for a in (
-            ('one@b.c', 'elegant'),
-            ('one@b.c', 'jspec'),
-            ('one@b.c', 'srw',),
-            ('one@b.c', 'synergia'),
-            ('one@b.c', 'warppba'),
-            ('one@b.c', 'warpvnd'),
-            ('two@b.c', 'elegant'),
-            ('two@b.c', 'jspec'),
-            ('two@b.c', 'srw',),
-            ('two@b.c', 'synergia'),
-            ('two@b.c', 'warppba'),
-            ('two@b.c', 'warpvnd'),
-            ('three@b.c', 'elegant'),
-            ('three@b.c', 'jspec'),
-            ('three@b.c', 'srw',),
-            ('three@b.c', 'synergia'),
-            ('three@b.c', 'warppba'),
-            ('three@b.c', 'warpvnd'),
-    ):
-        l.append(_run(*a))
-    await _cancel_on_exception(asyncio.gather(*l))
-
-
-async def _run_sequential_parallel(client):
-    c = []
-    s = CODES[client.sim_type]
-    e = s[random.randrange(len(s))]
-    random.shuffle(e.reports)
-    for r in e.reports:
-        c.append(client.sim_run(e.name, r))
-    await _cancel_on_exception(asyncio.gather(*c))
-
-
-@contextlib.contextmanager
-def _timer(description):
-    s = time.time()
-    yield
-    if 'run-status' not in description:
-        pkdlog('{} elapsed_time={}', description, time.time() - s)
+    s = _get_sims(
+        await _get_apps(
+            await _get_clients('one@b.c', 'two@b.c', 'three@b.c'),
+        ),
+    )
+    try:
+        t = asyncio.gather(*s)
```

Comment:
I would think we would want the signal handlers to be set before the gather, but maybe not. If a login fails, there's no server side state to manage.

_get_sims is the only code that needs to know about the gather, and therefore the code should be:

try:
    t = await _get_sims()
    _register_signal_handlers(t)
    await t

Reply:
Yes, there is no state to cleanup with login requests. Registering the signal handler after the login simplifies the handler code. If we registered it before the login we'd have to convert the login coroutines into tasks before we could cancel them. There is no upside to cancelling them so this seems like unnecessary confusion.

Yes and no, _get_sims() is the only code that needs to know about the gather but it isn't the only code that needs to know about the list of tasks that are going to be gathered. The exception handling code needs a reference to the list of tasks so it can cancel them. So I went with this:
```python
def _sims():
  # ...
  return (sims, asyncio.gather(*sims))

try:
  s, t = await _sims()
  _register_signal_handlers(t)
  await t
except Exception as e:
  await cancel_all_tasks(s)
  # ...
```

Having an await return another awaitable seems odd but maybe it is only odd to me because it is new...

Reply:
I don't think it is a problem that `_sims()` returns the tasks to wait on. I think the `await` on `t` is problematic since the `t` is passed to the signal handlers, too. Seems like that's the global state that needs to be managed, not sims.

I don't think `s` needs to be returned. Nobody uses it. There's a reference to `_sims` in the `except`, but I don't think that's correct or useful if it is `s`. There is a `cancel_all_tasks(s)`, but I think that should be `t`, not `s`.